File: common/src/main/java/de/m_marvin/industria/core/util/StructureFinder.java
package de.m_marvin.industria.core.util;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.function.Predicate;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.BlockState;

public class StructureFinder {
	
	public static Optional<List<BlockPos>> findStructure(Level level, BlockPos startPos, int maxBlocks, Predicate<BlockState> blockPredicate) {
		List<BlockPos> posList = new ArrayList<>();
		if (checkBlock(level, startPos, maxBlocks, posList, blockPredicate)) {
			return Optional.of(posList);
		}
		return Optional.empty();
	}
	
	protected static boolean checkBlock(Level level, BlockPos pos, int scanDepth, List<BlockPos> posList, Predicate<BlockState> blockPredicate) {
		if (!posList.contains(pos)) {
			BlockState state = level.getBlockState(pos);
			if (blockPredicate.test(state)) {
				posList.add(pos);
				if (scanDepth > 0) {
					for (Direction d : Direction.values()) {
						if (!checkBlock(level, pos.relative(d), scanDepth - 1, posList, blockPredicate)) {
							return false;
						}
					}
					return true;
				}
				return false;
			}
		}
		return true;
	}
	
}

File: common/src/main/kotlin/org/valkyrienskies/tournament/util/StructureFinder.kt
package org.valkyrienskies.tournament.util

import net.minecraft.core.BlockPos
import net.minecraft.server.level.ServerLevel
import org.valkyrienskies.core.impl.datastructures.DenseBlockPosSet
import org.valkyrienskies.mod.common.util.toJOML
import java.util.concurrent.atomic.AtomicInteger

object StructureFinder {

    private fun findStructureLoop(level: ServerLevel, pos: BlockPos, blacklist: Set<String>, set : DenseBlockPosSet, checked : MutableSet<BlockPos>, amount : AtomicInteger) {
        if(amount.get() > 2000)
            return

        if(checked.contains(pos))
            return

        checked.add(pos)

        if(blacklist.contains(level.getBlockState(pos).block.builtInRegistryHolder().key().toString()))
            return

        if(level.getBlockState(pos).isAir)
            return

        set.add(pos.toJOML())

        amount.getAndIncrement()

        findStructureLoop(level, BlockPos(pos.x+1, pos.y, pos.z), blacklist, set, checked, amount)
        findStructureLoop(level, BlockPos(pos.x, pos.y+1, pos.z), blacklist, set, checked, amount)
        findStructureLoop(level, BlockPos(pos.x, pos.y, pos.z+1), blacklist, set, checked, amount)

        findStructureLoop(level, BlockPos(pos.x-1, pos.y, pos.z), blacklist, set, checked, amount)
        findStructureLoop(level, BlockPos(pos.x, pos.y-1, pos.z), blacklist, set, checked, amount)
        findStructureLoop(level, BlockPos(pos.x, pos.y, pos.z-1), blacklist, set, checked, amount)
    }

    fun findStructure(level : ServerLevel, pos : BlockPos, blacklist : Set<String>) : DenseBlockPosSet {
        val set = DenseBlockPosSet()

        findStructureLoop(level, pos, blacklist, set, mutableSetOf(), AtomicInteger(0))

        return set
    }

}